[APM] Update python tracer logs instructions (#14344)
DataDog/documentation: content/en/tracing/troubleshooting/tracer_debug_logs.md

   By default, the file size is 15728640 bytes (about 15MB) and one backup log file is created. To increase the default log file size, specify the size in bytes with the `DD_TRACE_LOG_FILE_SIZE_BYTES` setting.

3. To route logs to the console, for **Python 2** applications, configure `logging.basicConfig()` or similar. Logs are automatically sent to the console for **Python 3** applications. 

### Scenario 2: ddtrace version 1.0.x to 1.2.x

1. To enable debug mode: `DD_TRACE_DEBUG=true`
    - `DD_CALL_BASIC_CONFIG=true`

2. To route logs to the console, for **Python 2 or Python 3** applications, configure logging.basicConfig() or use DD_CALL_BASIC_CONFIG=true.

### Scenario 3: ddtrace version 0.x

1. Set the `DD_TRACE_DEBUG` environment variable to `true` to enable debug mode.

2. To route logs to the console, for **Python 2 or Python 3** applications, configure logging.basicConfig() or use DD_CALL_BASIC_CONFIG=true.

### Scenario 4: Configuring debug logging in the application code with the standard logging library

> For any version of dd-trace-py, rather than setting the `DD_TRACE_DEBUG` tracer environment variable, you can enable debug logging in the application code by using the `logging` standard library directly:

Reviewer: ```suggestion
For any version of ddtrace, rather than setting the `DD_TRACE_DEBUG` tracer environment variable, you can enable debug logging in the application code by using the `logging` standard library directly:
```

{{< programming-lang lang="python" >}}

The steps for enabling debug mode Datadog Python Tracer, depends on which version of the tracer your application is using. Choose the scenario that applies:

### Scenario 1: dd-trace-py version 1.3.0 or higher

1. Set the `DD_TRACE_DEBUG` environment variable to `true` to enable debug mode.

> 2. To route debug logs to a log file, set `DD_TRACE_LOG_FILE` with a filename that tracer logs should be written to, for example, `DD_TRACE_LOG_FILE=ddtrace_logs.log`. 

Reviewer: Not sure if it's clear where on the filesystem `ddtrace_logs.log` will be? Is this relative to the current directory at the time the app is started?

Author: >Is this relative to the current directory at the time the app is started
By default, yes. This example will go to `<currentdirectory>/ddtrace_logs.log`.

To reference a different file, someone could do `/var/logs/ddtrace_logs.log` or similar. Should I add an example to route it elsewhere too to be more clear?

Reviewer: ```suggestion
2. To route debug logs to a log file, set `DD_TRACE_LOG_FILE` with a filename that tracer logs should be written to, relative to the current working directory. For example, `DD_TRACE_LOG_FILE=ddtrace_logs.log`. 
```

maybe something like this?

   By default, the file size is 15728640 bytes (about 15MB) and one backup log file is created. To increase the default log file size, specify the size in bytes with the `DD_TRACE_LOG_FILE_SIZE_BYTES` setting.

3. To route logs to the console, for **Python 2** applications, configure `logging.basicConfig()` or similar. Logs are automatically sent to the console for **Python 3** applications. 

### Scenario 2: dd-trace-py version 1.0.0 to 1.2.0

1. Set these environment variables to enable debug mode:
    - `DD_TRACE_DEBUG=true`
    - `DD_CALL_BASIC_CONFIG=true`

> 2. To route logs to the console, or **Python 2** applications, configure `logging.basicConfig()` or similar. Logs are automatically sent to the console for **Python 3** applications. 

Reviewer 2: Since you are explicitly saying to do `DD_CALL_BASIC_CONFIG=true` do they need this step?

Or maybe we shouldn't encourage the use of `DD_CALL_BASIC_CONFIG=true` but keep this step/statement and ask them to do so manually?

Author: For **Scenario 2: dd-trace-py version 1.0.0 to 1.2.0**, these two env vars came from our current docs, which tells people to set both right now: https://docs.datadoghq.com/tracing/troubleshooting/tracer_debug_logs/?code-lang=python, which states.

> To enable debug mode for the Datadog Python Tracer, set the environment variables DD_TRACE_DEBUG=true and DD_CALL_BASIC_CONFIG=true when using ddtrace-run.

In 1.x-1.2.x, setting `DD_TRACE_DEBUG=true` doesn't emit logs to the console unless `DD_CALL_BASIC_CONFIG=true` is used or `logging.basicConfig()` gets called.

What do you think of something like this to be clear why we're asking users to enable `DD_CALL_BASIC_CONFIG=true`?

> 1. To enable debug mode: `DD_TRACE_DEBUG=true`
> 2. To route logs to the console, for **Python 2** or **Python 3** applications, configure `logging.basicConfig()` or use `DD_CALL_BASIC_CONFIG=true`.

I mainly want to avoid the case where someone only has `DD_TRACE_DEBUG=true` configured, but don't have debug logs because `logging.basicConfig()` was never called.

Reviewer 2: yeah, I like the new suggestion. encourages `logging.basicConfig()` first. 👍🏻

### Scenario 3: dd-trace-py version 0.x

1. Set the `DD_TRACE_DEBUG` environment variable to `true` to enable debug mode.

> 2. To route logs to the console, for **Python 2** applications, configure `logging.basicConfig()` or similar. Logs are automatically sent to the console for **Python 3** applications. 

Reviewer 2: They shouldn't need this since we call basic config by default right?

Reviewer 2: we can probably just copy the same suggestion to have above to be consistent. what do you think?

Author: I updated both to read as 
> To route logs to the console, for **Python 2 or Python 3** applications, configure logging.basicConfig() or use DD_CALL_BASIC_CONFIG=true.

For 0.x, I think this is only needed if they use `patch_all/patch` but if a `ddtrace-run` application uses, I don't think it is a problem, so I guess this makes it easier to read through?

{{< programming-lang lang="python" >}}

> To enable debug mode for the Datadog Python Tracer, review the instructions below for the related scenario for the application being traced. Log emitted by the Python Tracer contain the name **[ddtrace]**.

Reviewer 3: ```suggestion
To enable debug mode for the Datadog Python Tracer, review the instructions below for the specific scenario that applies to the application being traced. Logs emitted by the Python Tracer contain the name **[ddtrace]**.
```

#### 1.3.0 or higher

> 

Reviewer 3: Why don't we mention `DD_TRACE_LOG_FILE_LEVEL` ? I actually don't totally understand why we'd have it in the first place, should we not add it in our corp docs?

Author: `DD_TRACE_LOG_FILE_LEVEL` doesn't need to be configured in order for logs to be written to a file. Since it wouldn't be part of the normal debug flow, I left it out of the instructions to keep it as simple as possible.

The PR https://github.com/DataDog/dd-trace-py/pull/3214 covers how `DD_TRACE_LOG_FILE_LEVEL` affects logging behavior when used with `DD_TRACE_DEBUG`. For anyone interested, it'll be in the API docs instead.

As for a use case of `DD_TRACE_LOG_FILE_LEVEL`, I think a good one would be if we want as little `DEBUG` logs as possible, but want to target exception logs more carefully, but that's still a rare scenario. I left it there as an option, but it'll be really rare to recommend it.

Reviewer 3: Thanks for the explanation! The plan seems good!

1. Set these environment variables to enable debug mode:
    - `DD_TRACE_DEBUG=true`

>     - `DD_TRACE_LOGGING_RATE=0` (Optional: used for more tracer logs)

Reviewer 2: ```suggestion
```

Not needed, this is now implied when setting `DD_TRACE_DEBUG=true`

Reviewer 2: Since 0.43.0

Author: Marking this as resolved !

2. To route to a log file:
    - Configure `DD_TRACE_LOG_FILE` with the log file that tracer logs should be written to. Example: `DD_TRACE_LOG_FILE=ddtrace_logs.log`.

>     - By default, the file size is `DD_TRACE_LOG_FILE_SIZE_BYTES=15728640` or ~15 mb and one backup log will be created. To increase the default log file size, change `DD_TRACE_LOG_FILE_SIZE_BYTES` to the desired file bytes size.

Reviewer 3: ```suggestion
    - By default, the file size is `DD_TRACE_LOG_FILE_SIZE_BYTES=15728640` or ~15 mb and one backup log file will be created. To increase the default log file size, change `DD_TRACE_LOG_FILE_SIZE_BYTES` to the desired file bytes size.
```
Just a specificity nitpick

3. To route logs to the console: 
    - This is done by default for **Python 3** applications. 
    - For **Python 2** applications, `logging.basicConfig()` or similar needs to be configured.

#### 1.0.0 to 1.2.0

1. Set these environment variables to enable debug mode:
    - `DD_TRACE_DEBUG=true`

>     - `DD_CALL_BASIC_CONFIG=true`

Reviewer 2: We don't want to encourage the usage of this env variable in 1.x, we should give the same guidance as above:

```
To route logs to the console: 
    - This is done by default for **Python 3** applications. 
    - For **Python 2** applications, `logging.basicConfig()` or similar needs to be configured.
```

>     - `DD_TRACE_LOGGING_RATE=0` (Optional: used for more tracer logs)

Reviewer 2: ```suggestion
```

2. `ddtrace` logs are routed to the console by default.

#### 0.x

1. Set these environment variables to enable debug mode::
    - `DD_TRACE_DEBUG=true`

>     - `DD_CALL_BASIC_CONFIG=true`

Reviewer 3: I think we don't actually need this envar since we use to call `basicConfig()` by default before 1.x https://github.com/DataDog/dd-trace-py/pull/3168

Reviewer 2: this is correct, for versions <1.0 we don't need to call `DD_CALL_BASIC_CONFIG=true`

>     - `DD_TRACE_LOGGING_RATE=0` (Optional: used for more tracer logs)

Reviewer 2: ```suggestion
```

Not needed since 0.43.0

{{< /programming-lang >}}
{{< programming-lang lang="python" >}}

> Logs generated by the Python Tracer have the logging handler name `[ddtrace]`. 

Reviewer 2: ```suggestion
Logs generated by the Python Tracer have the logging handler name `ddtrace`. 
```

It showing up as `[ddtrace]` is an implementation detail of the log format they are using

{{< /programming-lang >}}
{{< programming-lang lang="python" >}}

> Logs generated by the Python Tracer have the logging handler name [**ddtrace]**. 

Reviewer 3: We already said this on line 33, should we delete it from there or here? Not sure, just pointing it out. Feel free to resolve if you think it's good to have in both places.

Author: I added it in two places in case someone jumped straight to the "Review" section. Which place would be better to explain the name? In the "how to enable" portion or the "review" portion?
